README.md
# image_crop_widget

A Flutter widget to crop images. The widget is completely written in Dart and has minimal dependencies.

The widget displays the image within the given layout space. On top of the image a transparent rectangular overlay, with handles on each corner, is drawn. The overlay handles can be dragged by touch to adjust the cropping area.

![Example](https://github.com/flbaue/image_crop_widget/raw/master/assets/example.png)

By calling the `cropImage()` method on the widget's state object, the image area that is marked by the overlay is returned as a new image object.

To acquire the widget's state you can use a `GlobalKey` object.

Example:

```Dart
import 'dart:ui'; // This imports the 'Image' class.

final key = GlobalKey<ImageCropState>();
Image imageObject = ...

...

ImageCrop(key: key, image: imageObject) // This could be used inside a  build method.

...
await key.currentState.rotateImage(); // Rotate the image be 90° clockwise.
final cropedImage = await key.currentState.cropImage(); // This could be used inside a 'onPress' handler method.
```

## How to create an image object

The `Image` class from `dart:ui` is typically not instantiated directly. Instead, you could convert your image data into a `Uint8List` and instantiate the image like this:

```Dart
Uint8List bytes = ...
final codec = await instantiateImageCodec(bytes);
final frame = await codec.getNextFrame();
final image = frame.image;
```

## How to display an image object

The `Image` object can be displayed with the Flutter `Image` widget. One way to do this, is by converting the image into a `Uint8List` and pass it into the widget's memory constructor. Please note, that the widget is not the same `Image` class as the image object itself.

```Dart
final cropedImage = await key.currentState.cropImage();
final byte = await cropedImage.toByteData(format: ui.ImageByteFormat.png);
final byteList = byte.buffer.asUint8List();

Image.memory(byteList)
```

## How to persist an image object

The `Image` object can be persisted into a file or a database. To do this, you can convert the image object into a `Uint8List` and write it into a `File` or your database object.

```Dart
final cropedImage = await key.currentState.cropImage();
final byte = await cropedImage.toByteData(format: ui.ImageByteFormat.png);
final byteList = byte.buffer.asUint8List();

final imageFile = File("Some path and filename"); // You can use the path_provider package to locate the right path.
final result = await imageFile.writeAsBytes(byteList);
```


CHANGELOG.md
## [0.0.6] - 2019-11-01.

- Super smooth touch recognition.
- Refactoring

## [0.0.5] - 2019-10-30.

- Made the crop rect moveable.
- Corners will not jump anymore, when the touch event was not centered on them.

## [0.0.4] - 2019-10-08.

Added image rotation method.

## [0.0.3] - 2019-10-08.

Fixed the image link inside the readme.

## [0.0.2] - 2019-10-08.

Added a new example.

## [0.0.1] - 2019-10-08.

Initial release of the library.

- Includes a Flutter widget to display and crop an image.


test/src/crop_area_test.dart
// Copyright 2019 Florian Bauer. All rights reserved.
// Use of this source code is governed by a MIT license that can be
// found in the LICENSE file.

import 'dart:ui';

import 'package:flutter_test/flutter_test.dart';

import 'package:image_crop_widget/src/crop_area.dart';

void main() {
  test('moves top left corner and checks the bounds', () {
    final cropArea = CropArea(
      cornerSize: 10.0,
    );

    cropArea.initSizes(
      bounds: Rect.fromLTRB(10, 10, 190, 290),
      center: Offset(100.0, 150.0),
      width: 70.0,
      height: 80.0,
    );

    cropArea.moveTopLeftCorner(Offset(20.0, 20.0));
    expect(cropArea.cropRect.left, 20.0);
    expect(cropArea.cropRect.top, 20.0);
    expect(cropArea.cropRect.right, 135.0);
    expect(cropArea.cropRect.bottom, 190.0);

    cropArea.moveTopLeftCorner(Offset(0.0, 0.0));
    expect(cropArea.cropRect.left, 10.0);
    expect(cropArea.cropRect.top, 10.0);
    expect(cropArea.cropRect.right, 135.0);
    expect(cropArea.cropRect.bottom, 190.0);

    cropArea.moveTopLeftCorner(Offset(125.0, 180.0));
    expect(cropArea.cropRect.left, 125.0);
    expect(cropArea.cropRect.top, 180.0);
    expect(cropArea.cropRect.right, 135.0);
    expect(cropArea.cropRect.bottom, 190.0);

    cropArea.moveTopLeftCorner(Offset(130.0, 185.0));
    expect(cropArea.cropRect.left, 125.0);
    expect(cropArea.cropRect.top, 180.0);
    expect(cropArea.cropRect.right, 135.0);
    expect(cropArea.cropRect.bottom, 190.0);
  });

  test('moves top right corner and checks the bounds', () {
    final cropArea = CropArea(
      cornerSize: 10.0,
    );

    cropArea.initSizes(
      bounds: Rect.fromLTRB(10, 10, 190, 290),
      center: Offset(100.0, 150.0),
      width: 70.0,
      height: 80.0,
    );

    cropArea.moveTopRightCorner(Offset(180.0, 20.0));
    expect(cropArea.cropRect.left, 65.0);
    expect(cropArea.cropRect.top, 20.0);
    expect(cropArea.cropRect.right, 180.0);
    expect(cropArea.cropRect.bottom, 190.0);

    cropArea.moveTopRightCorner(Offset(200.0, 0.0));
    expect(cropArea.cropRect.left, 65.0);
    expect(cropArea.cropRect.top, 10.0);
    expect(cropArea.cropRect.right, 190.0);
    expect(cropArea.cropRect.bottom, 190.0);

    cropArea.moveTopRightCorner(Offset(0.0, 300.0));
    expect(cropArea.cropRect.left, 65.0);
    expect(cropArea.cropRect.top, 180.0);
    expect(cropArea.cropRect.right, 75.0);
    expect(cropArea.cropRect.bottom, 190.0);
  });

  test('moves bottom right corner and checks the bounds', () {
    final cropArea = CropArea(
      cornerSize: 10.0,
    );

    cropArea.initSizes(
      bounds: Rect.fromLTRB(10, 10, 190, 290),
      center: Offset(100.0, 150.0),
      width: 70.0,
      height: 80.0,
    );

    cropArea.moveBottomRightCorner(Offset(180.0, 280.0));
    expect(cropArea.cropRect.left, 65.0);
    expect(cropArea.cropRect.top, 110.0);
    expect(cropArea.cropRect.right, 180.0);
    expect(cropArea.cropRect.bottom, 280.0);

    cropArea.moveBottomRightCorner(Offset(200.0, 300.0));
    expect(cropArea.cropRect.left, 65.0);
    expect(cropArea.cropRect.top, 110.0);
    expect(cropArea.cropRect.right, 190.0);
    expect(cropArea.cropRect.bottom, 290.0);

    cropArea.moveBottomRightCorner(Offset(0.0, 0.0));
    expect(cropArea.cropRect.left, 65.0);
    expect(cropArea.cropRect.top, 110.0);
    expect(cropArea.cropRect.right, 75.0);
    expect(cropArea.cropRect.bottom, 120.0);
  });

  test('moves bottom left corner and checks the bounds', () {
    final cropArea = CropArea(
      cornerSize: 10.0,
    );

    cropArea.initSizes(
      bounds: Rect.fromLTRB(10, 10, 190, 290),
      center: Offset(100.0, 150.0),
      width: 70.0,
      height: 80.0,
    );

    cropArea.moveBottomLeftCorner(Offset(20.0, 280.0));
    expect(cropArea.cropRect.left, 20.0);
    expect(cropArea.cropRect.top, 110.0);
    expect(cropArea.cropRect.right, 135.0);
    expect(cropArea.cropRect.bottom, 280.0);

    cropArea.moveBottomLeftCorner(Offset(0.0, 300.0));
    expect(cropArea.cropRect.left, 10.0);
    expect(cropArea.cropRect.top, 110.0);
    expect(cropArea.cropRect.right, 135.0);
    expect(cropArea.cropRect.bottom, 290.0);

    cropArea.moveBottomLeftCorner(Offset(200.0, 0.0));
    expect(cropArea.cropRect.left, 125.0);
    expect(cropArea.cropRect.top, 110.0);
    expect(cropArea.cropRect.right, 135.0);
    expect(cropArea.cropRect.bottom, 120.0);
  });

  test('moves a crop area and checks the bounds', () {
    final cropArea = CropArea(
      cornerSize: 15.0,
    );

    cropArea.initSizes(
      bounds: Rect.fromLTRB(10, 10, 190, 290),
      center: Offset(100.0, 150.0),
      width: 70.0,
      height: 80.0,
    );

    cropArea.moveArea(Offset(80.0, 80.0));
    expect(cropArea.cropRect.left, 45.0);
    expect(cropArea.cropRect.top, 40.0);
    expect(cropArea.cropRect.right, 115.0);
    expect(cropArea.cropRect.bottom, 120.0);

    cropArea.moveArea(Offset(20.0, 20.0));
    expect(cropArea.cropRect.left, 10.0);
    expect(cropArea.cropRect.top, 10.0);
    expect(cropArea.cropRect.right, 80.0);
    expect(cropArea.cropRect.bottom, 90.0);

    cropArea.moveArea(Offset(180.0, 280.0));
    expect(cropArea.cropRect.left, 120.0);
    expect(cropArea.cropRect.top, 210.0);
    expect(cropArea.cropRect.right, 190.0);
    expect(cropArea.cropRect.bottom, 290.0);
  });

  test('creates a crop area and checks the contains method', () {
    final cropArea = CropArea(
      cornerSize: 15.0,
    );

    cropArea.initSizes(
      bounds: Rect.fromLTRB(10, 10, 190, 290),
      center: Offset(100.0, 150.0),
      width: 70.0,
      height: 80.0,
    );

    expect(
      cropArea.contains(Offset(65.0, 110.0)),
      true,
      reason: "top left corner",
    );
    expect(
      cropArea.contains(Offset(135.0, 110.0)),
      true,
      reason: "top right corner",
    );
    expect(
      cropArea.contains(Offset(135.0, 190.0)),
      true,
      reason: "bottom right corner",
    );
    expect(
      cropArea.contains(Offset(65.0, 190.0)),
      true,
      reason: "bottom left corner",
    );

    expect(
      cropArea.contains(Offset(57.5, 102.5)),
      true,
      reason: "top left corner of corner rect",
    );
    expect(
      cropArea.contains(Offset(57.4, 102.4)),
      false,
      reason: "outside top left corner of corner rect",
    );

    expect(
      cropArea.contains(Offset(142.4, 102.6)),
      true,
      reason: "top right corner of corner rect",
    );
    expect(
      cropArea.contains(Offset(142.5, 102.5)),
      false,
      reason: "top right corner of corner rect",
    );

    expect(
      cropArea.contains(Offset(142.4, 197.4)),
      true,
      reason: "bottom right corner of corner rect",
    );
    expect(
      cropArea.contains(Offset(142.5, 197.5)),
      false,
      reason: "bottom right corner of corner rect",
    );

    expect(
      cropArea.contains(Offset(57.5, 197.4)),
      true,
      reason: "bottom left corner of corner rect",
    );
    expect(
      cropArea.contains(Offset(57.4, 197.5)),
      false,
      reason: "bottom left corner of corner rect",
    );
  });
}


test/image_crop_widget_test.dart
// Copyright 2019 Florian Bauer. All rights reserved.
// Use of this source code is governed by a MIT license that can be
// found in the LICENSE file.

import 'package:flutter_test/flutter_test.dart';

import 'package:image_crop_widget/image_crop_widget.dart';

// void main() {
//   test('adds one to input values', () {
//     final calculator = Calculator();
//     expect(calculator.addOne(2), 3);
//     expect(calculator.addOne(-7), -6);
//     expect(calculator.addOne(0), 1);
//     expect(() => calculator.addOne(null), throwsNoSuchMethodError);
//   });
// }


lib/src/crop_area.dart
// Copyright 2019 Florian Bauer. All rights reserved.
// Use of this source code is governed by a MIT license that can be
// found in the LICENSE file.

import 'dart:math';
import 'dart:ui';

import 'package:flutter/foundation.dart';

class CropArea {
  final double _cornerSize;
  Rect _bounds;
  Rect _cropRect;
  Rect _topLeftCorner;
  Rect _topRightCorner;
  Rect _bottomRightCorner;
  Rect _bottomLeftCorner;

  CropArea({
    double cornerSize = 32.0,
  }) : _cornerSize = cornerSize;

  void initSizes({
    @required Offset center,
    @required double width,
    @required double height,
    @required Rect bounds,
  }) {
    _bounds = bounds;
    _cropRect = Rect.fromCenter(center: center, width: width, height: height);
    _updateCorners();
  }

  Rect get cropRect => _cropRect;

  void _updateCorners() {
    _topLeftCorner = Rect.fromCenter(
      center: _cropRect.topLeft,
      width: _cornerSize,
      height: _cornerSize,
    );
    _topRightCorner = Rect.fromCenter(
      center: _cropRect.topRight,
      width: _cornerSize,
      height: _cornerSize,
    );
    _bottomRightCorner = Rect.fromCenter(
      center: _cropRect.bottomRight,
      width: _cornerSize,
      height: _cornerSize,
    );
    _bottomLeftCorner = Rect.fromCenter(
      center: _cropRect.bottomLeft,
      width: _cornerSize,
      height: _cornerSize,
    );
  }

  bool contains(Offset position) {
    return _cropRect.contains(position) ||
        _topLeftCorner.contains(position) ||
        _topRightCorner.contains(position) ||
        _bottomRightCorner.contains(position) ||
        _bottomLeftCorner.contains(position);
  }

  void moveArea(Offset newCenter) {
    final newRect = Rect.fromCenter(
      center: newCenter,
      width: _cropRect.width,
      height: _cropRect.height,
    );

    var offset = Offset(0.0, 0.0);
    if (newRect.left < _bounds.left) {
      offset = offset.translate(_bounds.left - newRect.left, 0.0);
    }

    if (newRect.top < _bounds.top) {
      offset = offset.translate(0.0, _bounds.top - newRect.top);
    }

    if (newRect.right > _bounds.right) {
      offset = offset.translate(_bounds.right - newRect.right, 0.0);
    }

    if (newRect.bottom > _bounds.bottom) {
      offset = offset.translate(0.0, _bounds.bottom - newRect.bottom);
    }

    _cropRect = newRect.shift(offset);
    _updateCorners();
  }

  double _applyLeftBounds(double left) {
    var boundedLeft = max(
      left,
      _bounds.left,
    ); // left bound
    boundedLeft = min(
      boundedLeft,
      _cropRect.right - _cornerSize,
    ); // right bound
    return boundedLeft;
  }

  double _applyTopBounds(double top) {
    var boundedTop = max(
      top,
      _bounds.top,
    ); // top bound
    boundedTop = min(
      boundedTop,
      _cropRect.bottom - _cornerSize,
    ); // bottom bound
    return boundedTop;
  }

  double _applyRightBounds(double right) {
    var boundedRight = min(
      right,
      _bounds.right,
    ); // right bound
    boundedRight = max(
      boundedRight,
      _cropRect.left + _cornerSize,
    ); // left bound
    return boundedRight;
  }

  double _applyBottomBounds(double bottom) {
    var boundedBottom = min(
      bottom,
      _bounds.bottom,
    ); // bottom bound
    boundedBottom = max(
      boundedBottom,
      _cropRect.top + _cornerSize,
    ); // top bound
    return boundedBottom;
  }

  void moveTopLeftCorner(Offset newPosition) {
    _cropRect = Rect.fromLTRB(
      _applyLeftBounds(newPosition.dx),
      _applyTopBounds(newPosition.dy),
      _cropRect.right,
      _cropRect.bottom,
    );
    _updateCorners();
  }

  void moveTopRightCorner(Offset newPosition) {
    _cropRect = Rect.fromLTRB(
      _cropRect.left,
      _applyTopBounds(newPosition.dy),
      _applyRightBounds(newPosition.dx),
      _cropRect.bottom,
    );
    _updateCorners();
  }

  void moveBottomRightCorner(Offset newPosition) {
    _cropRect = Rect.fromLTRB(
      _cropRect.left,
      _cropRect.top,
      _applyRightBounds(newPosition.dx),
      _applyBottomBounds(newPosition.dy),
    );
    _updateCorners();
  }

  void moveBottomLeftCorner(Offset newPosition) {
    _cropRect = Rect.fromLTRB(
      _applyLeftBounds(newPosition.dx),
      _cropRect.top,
      _cropRect.right,
      _applyBottomBounds(newPosition.dy),
    );
    _updateCorners();
  }

  Offset _getTopLeftDelta(Offset position) {
    return _topLeftCorner.center - position;
  }

  Offset _getTopRightDelta(Offset position) {
    return _topRightCorner.center - position;
  }

  Offset _getBottomRightDelta(Offset position) {
    return _bottomRightCorner.center - position;
  }

  Offset _getBottomLeftDelta(Offset position) {
    return _bottomLeftCorner.center - position;
  }

  Offset _getCenterDelta(Offset position) {
    return _cropRect.center - position;
  }

  CropActionArea getActionArea(Offset position) {
    if (_topLeftCorner.contains(position)) {
      return CropActionArea.top_left;
    }

    if (_topRightCorner.contains(position)) {
      return CropActionArea.top_right;
    }

    if (_bottomLeftCorner.contains(position)) {
      return CropActionArea.bottom_left;
    }

    if (_bottomRightCorner.contains(position)) {
      return CropActionArea.bottom_right;
    }

    if (_cropRect.contains(position)) {
      return CropActionArea.center;
    }

    return CropActionArea.none;
  }

  Offset getActionAreaDelta(Offset position, CropActionArea activeArea) {
    switch (activeArea) {
      case CropActionArea.top_left:
        return _getTopLeftDelta(position);
      case CropActionArea.top_right:
        return _getTopRightDelta(position);
      case CropActionArea.bottom_right:
        return _getBottomRightDelta(position);
      case CropActionArea.bottom_left:
        return _getBottomLeftDelta(position);
      case CropActionArea.center:
        return _getCenterDelta(position);
      default:
        return Offset.zero;
    }
  }

  void move(Offset position, Offset delta, CropActionArea actionArea) {
    switch (actionArea) {
      case CropActionArea.top_left:
        moveTopLeftCorner(position + delta);
        break;
      case CropActionArea.top_right:
        moveTopRightCorner(position + delta);
        break;
      case CropActionArea.bottom_right:
        moveBottomRightCorner(position + delta);
        break;
      case CropActionArea.bottom_left:
        moveBottomLeftCorner(position + delta);
        break;
      case CropActionArea.center:
        moveArea(position + delta);
        break;
      default:
    }
  }
}

enum CropActionArea {
  top_left,
  top_right,
  bottom_right,
  bottom_left,
  center,
  none
}


example/example_screen.dart
import 'dart:ui' as ui;

import 'package:flutter/material.dart';
import 'package:image_crop_widget/image_crop_widget.dart';

class ExampleScreen extends StatefulWidget {
  final ui.Image imageFile;
  ExampleScreen(this.imageFile);

  @override
  _ExampleScreenState createState() => _ExampleScreenState();
}

class _ExampleScreenState extends State<ExampleScreen> {
  final key = GlobalKey<ImageCropState>();

  void _onCropPress() async {
    final croppedImage = await key.currentState.cropImage();

    showModalBottomSheet(
      context: context,
      builder: (context) => RawImage(
        image: croppedImage,
        fit: BoxFit.contain,
        height: croppedImage.height.toDouble(),
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SafeArea(
        child: Column(
          children: <Widget>[
            Expanded(
              child: ImageCrop(key: key, image: widget.imageFile),
            ),
            FlatButton(
              child: Text("Okay"),
              onPressed: _onCropPress,
            ),
          ],
        ),
      ),
    );
  }
}


lib/src/pan_gesture_recognizer.dart
import 'package:flutter/foundation.dart';
import 'package:flutter/gestures.dart';

typedef OnPanUpdate = Function(PointerEvent);
typedef OnPanEnd = Function(PointerEvent);

class PanGestureRecognizer extends OneSequenceGestureRecognizer {
  final OnPanUpdate onPanUpdate;
  final OnPanEnd onPanEnd;

  PanGestureRecognizer({@required this.onPanUpdate, @required this.onPanEnd});

  @override
  void addPointer(PointerEvent event) {
    startTrackingPointer(event.pointer);
    resolve(GestureDisposition.accepted);
  }

  @override
  void handleEvent(PointerEvent event) {
    if (event is PointerMoveEvent) {
      onPanUpdate(event);
    } else if (event is PointerDownEvent) {
      onPanUpdate(event);
    } else if (event is PointerUpEvent) {
      onPanEnd(event);
      stopTrackingPointer(event.pointer);
    } else {
      print("PanGestureRecognizer Pointer Event: ${event.runtimeType}");
    }
  }

  @override
  String get debugDescription => 'PanGestureRecognizer';

  @override
  void didStopTrackingLastPointer(int pointer) {}
}


lib/src/crop_area_touch_handler.dart
// Copyright 2019 Florian Bauer. All rights reserved.
// Use of this source code is governed by a MIT license that can be
// found in the LICENSE file.

import 'dart:ui';

import 'package:flutter/foundation.dart';

import 'crop_area.dart';

class CropAreaTouchHandler {
  final CropArea _cropArea;
  Offset _activeAreaDelta;
  CropActionArea _activeArea;

  CropAreaTouchHandler({@required CropArea cropArea}) : _cropArea = cropArea;

  void startTouch(Offset touchPosition) {
    _activeArea = _cropArea.getActionArea(touchPosition);
    _activeAreaDelta = _cropArea.getActionAreaDelta(touchPosition, _activeArea);
  }

  void updateTouch(Offset touchPosition) {
    _cropArea.move(touchPosition, _activeAreaDelta, _activeArea);
  }

  void endTouch() {
    _activeArea = null;
    _activeAreaDelta = null;
  }
}


lib/image_crop_widget.dart
// Copyright 2019 Florian Bauer. All rights reserved.
// Use of this source code is governed by a MIT license that can be
// found in the LICENSE file.

library image_crop_widget;

import 'dart:math';
import 'dart:ui' as ui;
import 'package:flutter/material.dart';

import 'src/crop_area.dart';
import 'src/crop_area_touch_handler.dart';
import 'src/pan_gesture_recognizer.dart';

class ImageCrop extends StatefulWidget {
  final ui.Image image;

  ImageCrop({Key key, this.image})
      : assert(image != null),
        super(key: key);

  @override
  ImageCropState createState() => ImageCropState();
}

class ImageCropState extends State<ImageCrop> {
  /// Rotates the image clockwise by 90 degree.
  /// Completes when the rotation is done.
  Future<void> rotateImage() async {
    final image = _state.image;
    ui.Image newImage;
    var pixel = 0;
    var attempts = 0;

    /// This loop is a very hacky workaround.
    /// In on device tests, about 1 in 10 times, the picture.toImage method returned
    /// a corrupted image object. So far, it is unclear why this happens.
    ///
    /// In case the image is corrupted, two possibilities have been observed:
    /// 1. All bytes of the image are 0.
    /// 2. An index out of bounds exception was thrown when accessing a high index.
    ///
    /// Usually, this behavior was observed on the first attempt to call this rotateImage method
    /// after the widget was build. Following calls to this method never or rarely produced this issue.
    do {
      final recorder = ui.PictureRecorder();
      final canvas = Canvas(recorder);

      canvas.rotate(pi / 2.0);
      canvas.translate(0.0, -image.height.toDouble());
      canvas.drawImage(image, Offset.zero, Paint());

      final picture = recorder.endRecording();
      newImage = await picture.toImage(image.height, image.width);
      picture.dispose();

      final newByteData = await newImage.toByteData();
      try {
        pixel = newByteData.getUint8(newByteData.lengthInBytes - 1);
      } catch (e) {
        print(e);
      }

      attempts++;
    } while (pixel == 0 && attempts < 4);

    setState(() {
      _state.image = newImage;
    });
  }

  /// Crops the image to the currently marked area.
  /// Returns a new [ui.Image].
  Future<ui.Image> cropImage() async {
    final yOffset =
        (_state.widgetSize.height - _state.fittedImageSize.destination.height) /
            2.0;
    final xOffset =
        (_state.widgetSize.width - _state.fittedImageSize.destination.width) /
            2.0;
    final fittedCropRect = Rect.fromCenter(
      center: Offset(
        _state.cropArea.cropRect.center.dx - xOffset,
        _state.cropArea.cropRect.center.dy - yOffset,
      ),
      width: _state.cropArea.cropRect.width,
      height: _state.cropArea.cropRect.height,
    );

    final scale =
        _state.imageSize.width / _state.fittedImageSize.destination.width;
    final imageCropRect = Rect.fromLTRB(
        fittedCropRect.left * scale,
        fittedCropRect.top * scale,
        fittedCropRect.right * scale,
        fittedCropRect.bottom * scale);

    final recorder = ui.PictureRecorder();
    final canvas = Canvas(recorder);

    canvas.drawImage(
      _state.image,
      Offset(-imageCropRect.left, -imageCropRect.top),
      Paint(),
    );

    final picture = recorder.endRecording();
    final croppedImage = await picture.toImage(
      imageCropRect.width.toInt(),
      imageCropRect.height.toInt(),
    );
    picture.dispose();

    return croppedImage;
  }

  _SharedCropState _state = _SharedCropState();

  @override
  void initState() {
    super.initState();
    _state.image = widget.image;
    _state.cropArea = CropArea();
    _state.cropAreaTouchHandler =
        CropAreaTouchHandler(cropArea: _state.cropArea);
  }

  void _onPanUpdate(PointerEvent event) {
    _onUpdate(event.position);
  }

  void _onPanEnd(PointerEvent event) {
    setState(() {
      _state.lastTouchPosition = null;
      _state.touchPosition = null;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      width: double.infinity,
      height: double.infinity,
      color: Colors.black87,
      child: CustomPaint(
        painter: _ImagePainter(_state),
        foregroundPainter: _OverlayPainter(_state),
        child: RawGestureDetector(
          gestures: {
            PanGestureRecognizer:
                GestureRecognizerFactoryWithHandlers<PanGestureRecognizer>(
              () => PanGestureRecognizer(
                onPanUpdate: _onPanUpdate,
                onPanEnd: _onPanEnd,
              ),
              (PanGestureRecognizer instance) {},
            )
          },
        ),
      ),
    );
  }

  void _onUpdate(Offset globalPosition) {
    final RenderBox renderBox = context.findRenderObject();
    _state.lastTouchPosition = _state.touchPosition;
    _state.touchPosition = renderBox.globalToLocal(globalPosition);

    if (_state.lastTouchPosition == null) {
      _state.cropAreaTouchHandler.startTouch(_state.touchPosition);
    } else {
      _state.cropAreaTouchHandler.updateTouch(_state.touchPosition);
    }

    setState(() {});
  }
}

class _SharedCropState {
  ui.Image image;

  Offset touchPosition;
  Offset lastTouchPosition;

  Size widgetSize;
  Size imageSize;
  FittedSizes fittedImageSize;
  double horizontalSpacing;
  double verticalSpacing;
  Rect imageContainingRect;

  CropArea cropArea;
  CropAreaTouchHandler cropAreaTouchHandler;
}

class _ImagePainter extends CustomPainter {
  final _SharedCropState state;
  final ui.Image image;

  _ImagePainter(this.state) : image = state.image;

  @override
  void paint(ui.Canvas canvas, ui.Size size) {
    final displayRect = Rect.fromLTWH(0.0, 0.0, size.width, size.height);
    state.widgetSize = size;
    paintImage(
      canvas: canvas,
      image: state.image,
      rect: displayRect,
      fit: BoxFit.contain,
    );
    state.imageSize = Size(
      state.image.width.toDouble(),
      state.image.height.toDouble(),
    );
    state.fittedImageSize = applyBoxFit(
      BoxFit.contain,
      state.imageSize,
      size,
    );
    state.horizontalSpacing =
        (state.widgetSize.width - state.fittedImageSize.destination.width) / 2;
    state.verticalSpacing =
        (state.widgetSize.height - state.fittedImageSize.destination.height) /
            2;
    state.imageContainingRect = Rect.fromLTWH(
        state.horizontalSpacing,
        state.verticalSpacing,
        state.fittedImageSize.destination.width,
        state.fittedImageSize.destination.height);
  }

  @override
  bool shouldRepaint(_ImagePainter oldDelegate) {
    return image != oldDelegate.image;
  }
}

class _OverlayPainter extends CustomPainter {
  final _SharedCropState _state;
  final Rect _cropRect;
  final paintCorner = Paint()
    ..strokeWidth = 10.0
    ..strokeCap = StrokeCap.round
    ..color = Colors.white;
  final paintBackground = Paint()..color = Colors.white30;

  _OverlayPainter(this._state) : _cropRect = _state.cropArea.cropRect;

  @override
  void paint(Canvas canvas, Size size) {
    if (_state.cropArea.cropRect == null) {
      _state.cropArea.initSizes(
        bounds: _state.imageContainingRect,
        center: Offset(size.width / 2, size.height / 2),
        height: 100.0,
        width: 100.0,
      );
    }

    canvas.drawRect(_state.cropArea.cropRect, paintBackground);
    final points = <Offset>[
      _state.cropArea.cropRect.topLeft,
      _state.cropArea.cropRect.topRight,
      _state.cropArea.cropRect.bottomLeft,
      _state.cropArea.cropRect.bottomRight
    ];

    canvas.drawPoints(ui.PointMode.points, points, paintCorner);
  }

  @override
  bool shouldRepaint(_OverlayPainter oldDelegate) {
    return _cropRect != oldDelegate._cropRect;
  }
}
